apps/managedidentity/managedidentity_test.go

```diff
@@ -4,43 +4,319 @@ package managedidentity
+				t.Fatalf("testManagedIdentity: accesstoken should be nil")
+			}
+		})
+	}
+}
 
-	if err != nil {
-		t.Fatal(err)
+func Test_SystemAssigned_Returns_Token_Success(t *testing.T) {
+	testCases := []resourceTestData{
+		{source: DefaultToIMDS, endpoint: imdsEndpoint, resource: resource},
+		{source: DefaultToIMDS, endpoint: imdsEndpoint, resource: resourceDefaultSuffix},
 	}
+	for _, testCase := range testCases {
 
-	_, err = client.AcquireToken(context.Background(), "scope", WithClaims("claim"))
+		t.Run(string(testCase.source), func(t *testing.T) {
+			var url string
+			mockClient := mock.Client{}
+			responseBody, err := getSuccessfulResponse(resource)
+			if err != nil {
+				t.Fatalf("TestManagedIdentity: Error while forming json response : %s", err.Error())
+			}
+			mockClient.AppendResponse(mock.WithHTTPStatusCode(http.StatusOK), mock.WithBody(responseBody), mock.WithCallback(func(r *http.Request) { url = r.URL.String() }))
+			client, err := New(SystemAssigned(), WithHTTPClient(&mockClient))
 
-	if err == nil {
-		t.Errorf("TestManagedIdentity: unexpected nil error from TestManagedIdentity")
+			if err != nil {
+				t.Fatal(err)
+			}
+			result, err := client.AcquireToken(context.Background(), testCase.resource)
+			if !strings.HasPrefix(url, testCase.endpoint) {
+				t.Fatalf("TestManagedIdentity: URL request is not on %s fgot %s", testCase.endpoint, url)
+			}
+			if err != nil {
+				t.Fatalf("TestManagedIdentity: unexpected nil error from TestManagedIdentity %s", err.Error())
+			}
+			if result.AccessToken != token {
+				t.Fatalf("TestManagedIdentity: wanted %q, got %q", token, result.AccessToken)
+			}
+		})
 	}
 }
 
-func TestManagedIdentityWithClaims(t *testing.T) {
-	client, err := fakeClient(ClientID("123"))
+func TestCreatingIMDSClient(t *testing.T) {
+	tests := []struct {
+		name    string
+		id      ID
+		wantErr bool
+	}{
+		{
+			name: "System Assigned",
+			id:   SystemAssigned(),
+		},
+		{
+			name: "Client ID",
+			id:   ClientID("test-client-id"),
+		},
+		{
+			name: "Resource ID",
+			id:   ResourceID("test-resource-id"),
+		},
+		{
+			name: "Object ID",
+			id:   ObjectID("test-object-id"),
+		},
+		{
+			name:    "Empty Client ID",
+			id:      ClientID(""),
+			wantErr: true,
+		},
+		{
+			name:    "Empty Resource ID",
+			id:      ResourceID(""),
+			wantErr: true,
+		},
+		{
+			name:    "Empty Object ID",
+			id:      ObjectID(""),
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client, err := New(tt.id)
+			if tt.wantErr {
+				if err == nil {
+					t.Fatal("client New() should return a error but did not.")
+				}
+				return
+			}
+			if err != nil {
+				t.Fatal("client New() error while creating client")
+			} else {
+				if client.miType.value() != tt.id.value() {
+					t.Fatal("client New() did not assign a correct value to type.")
+				}
+			}
+		})
+	}
 
-	if err != nil {
-		t.Fatal(err)
+}
+func TestCreateIMDSAuthRequest(t *testing.T) {
```

Comment:
Suggest merging this into `Test_SystemAssigned_Returns_Token_Success`. Driving tests through the public API is generally better--what happens if we refactor `createIMDSAuthRequest` out of existence or `AcquireToken` stops calling it?--and that test already has a callback validating the request URL.

Reply:
Updated and now only tests are for public api and we check for every step, even for creating of request and return a fatal error